File: Android/app/src/main/java/com/neoremote/android/core/touch/TouchSurfaceInputAdapter.kt
package com.neoremote.android.core.touch

import com.neoremote.android.core.model.DragState
import com.neoremote.android.core.model.MouseButtonKind
import com.neoremote.android.core.model.RemoteCommand
import com.neoremote.android.core.model.TouchPoint
import com.neoremote.android.core.model.TouchSensitivitySettings
import com.neoremote.android.core.model.TouchSurfaceOutput
import com.neoremote.android.core.model.TouchSurfaceSemanticEvent
import kotlin.math.abs
import kotlin.math.hypot

class TouchSurfaceInputAdapter(
    settings: TouchSensitivitySettings = TouchSensitivitySettings(),
    private val tapDistanceThreshold: Float = 12f,
    private val tapDurationThreshold: Double = 0.22,
    private val doubleTapWindow: Double = 0.32,
    private val dragDistanceThreshold: Float = 14f,
    private val rightDragHoldDelay: Double = 0.18,
    private val scrollDominanceThreshold: Float = 1.15f,
) {
    private var moveMultiplier: Double = 1.0
    private var scrollMultiplier: Double = 1.0
    private var scrollActivationDistance: Float = 14f

    private data class ActiveTouch(
        val startPoint: TouchPoint,
        val point: TouchPoint,
        val startTime: Double,
    )

    private enum class Phase {
        IDLE,
        SINGLE_TAP_CANDIDATE,
        LEFT_DRAG_CANDIDATE,
        LEFT_DRAG_ACTIVE,
        MULTI_TOUCH_CANDIDATE,
        SCROLL_ACTIVE,
        RIGHT_DRAG_ACTIVE,
        MIDDLE_TAP_CANDIDATE,
    }

    private val activeTouches = linkedMapOf<Int, ActiveTouch>()
    private var phase = Phase.IDLE
    private var lastTapTime: Double? = null
    private var sessionStartTime: Double? = null
    private var sessionStartCentroid: TouchPoint? = null
    private var previousCentroid: TouchPoint? = null
    private var maxTouchCount = 0
    private var singleTouchId: Int? = null

    init {
        apply(settings)
    }

    fun apply(settings: TouchSensitivitySettings) {
        val clamped = settings.clamped
        moveMultiplier = clamped.cursorSensitivity
        scrollMultiplier = clamped.swipeSensitivity
        scrollActivationDistance = (14.0 / clamped.swipeSensitivity).toFloat()
    }

    fun touchBegan(id: Int, point: TouchPoint, timestamp: Double): TouchSurfaceOutput {
        if (activeTouches.isEmpty()) {
            beginSession(id = id, point = point, timestamp = timestamp)
        } else {
            activeTouches[id] = ActiveTouch(startPoint = point, point = point, startTime = timestamp)
            maxTouchCount = maxOf(maxTouchCount, activeTouches.size)
            sessionStartCentroid = centroid()
            previousCentroid = sessionStartCentroid
            singleTouchId = null

            phase = when (activeTouches.size) {
                2 -> Phase.MULTI_TOUCH_CANDIDATE
                3 -> Phase.MIDDLE_TAP_CANDIDATE
                else -> Phase.IDLE
            }
        }

        return TouchSurfaceOutput()
    }

    fun touchMoved(id: Int, point: TouchPoint, timestamp: Double): TouchSurfaceOutput {
        val existing = activeTouches[id] ?: return TouchSurfaceOutput()
        val previousPoint = existing.point
        val oldCentroid = previousCentroid ?: centroid()

        activeTouches[id] = existing.copy(point = point)

        return when (phase) {
            Phase.SINGLE_TAP_CANDIDATE -> handleSingleMove(from = previousPoint, to = point)
            Phase.LEFT_DRAG_CANDIDATE -> handleLeftDragCandidateMove(id = id, point = point)
            Phase.LEFT_DRAG_ACTIVE -> {
                val dx = (point.x - previousPoint.x) * moveMultiplier
                val dy = (point.y - previousPoint.y) * moveMultiplier
                TouchSurfaceOutput(
                    commands = listOf(RemoteCommand.Drag(DragState.CHANGED, dx, dy, MouseButtonKind.PRIMARY)),
                    semanticEvent = TouchSurfaceSemanticEvent.PRIMARY_DRAG_CHANGED,
                )
            }

            Phase.MULTI_TOUCH_CANDIDATE,
            Phase.SCROLL_ACTIVE,
            Phase.RIGHT_DRAG_ACTIVE,
            -> handleTwoFingerMove(timestamp = timestamp, oldCentroid = oldCentroid)

            Phase.MIDDLE_TAP_CANDIDATE -> {
                val start = sessionStartCentroid
                if (start != null && start.distanceTo(centroid()) > tapDistanceThreshold) {
                    phase = Phase.IDLE
                }
                TouchSurfaceOutput()
            }

            Phase.IDLE -> TouchSurfaceOutput()
        }
    }

    fun touchEnded(id: Int, point: TouchPoint, timestamp: Double): TouchSurfaceOutput {
        if (!activeTouches.containsKey(id)) return TouchSurfaceOutput()
        activeTouches[id] = activeTouches.getValue(id).copy(point = point)

        val dragEnd = dragEndOutputIfNeeded(endingTouchId = id)
        activeTouches.remove(id)

        if (dragEnd != null) {
            resetIfSessionEnded()
            return dragEnd
        }

        if (activeTouches.isNotEmpty()) return TouchSurfaceOutput()

        val duration = timestamp - (sessionStartTime ?: timestamp)
        val movement = sessionStartCentroid?.let { start ->
            start.distanceTo(previousCentroid ?: start)
        } ?: 0f
        val output = when (phase) {
            Phase.SINGLE_TAP_CANDIDATE,
            Phase.LEFT_DRAG_CANDIDATE,
            -> {
                if (duration <= tapDurationThreshold && movement <= tapDistanceThreshold) {
                    lastTapTime = timestamp
                    TouchSurfaceOutput(
                        commands = listOf(RemoteCommand.Tap(MouseButtonKind.PRIMARY)),
                        semanticEvent = TouchSurfaceSemanticEvent.PRIMARY_TAP,
                    )
                } else {
                    TouchSurfaceOutput()
                }
            }

            Phase.MULTI_TOUCH_CANDIDATE -> {
                if (maxTouchCount == 2 && duration <= tapDurationThreshold && movement <= tapDistanceThreshold) {
                    TouchSurfaceOutput(
                        commands = listOf(RemoteCommand.Tap(MouseButtonKind.SECONDARY)),
                        semanticEvent = TouchSurfaceSemanticEvent.SECONDARY_TAP,
                    )
                } else {
                    TouchSurfaceOutput()
                }
            }

            Phase.MIDDLE_TAP_CANDIDATE -> {
                if (maxTouchCount == 3 && duration <= tapDurationThreshold && movement <= tapDistanceThreshold) {
                    TouchSurfaceOutput(
                        commands = listOf(RemoteCommand.Tap(MouseButtonKind.MIDDLE)),
                        semanticEvent = TouchSurfaceSemanticEvent.MIDDLE_TAP,
                    )
                } else {
                    TouchSurfaceOutput()
                }
            }

            else -> TouchSurfaceOutput()
        }

        resetSession()
        return output
    }

    fun cancelAllTouches(): TouchSurfaceOutput {
        val output = when (phase) {
            Phase.LEFT_DRAG_ACTIVE -> TouchSurfaceOutput(
                commands = listOf(RemoteCommand.Drag(DragState.ENDED, 0.0, 0.0, MouseButtonKind.PRIMARY)),
                semanticEvent = TouchSurfaceSemanticEvent.PRIMARY_DRAG_ENDED,
            )

            Phase.RIGHT_DRAG_ACTIVE -> TouchSurfaceOutput(
                commands = listOf(RemoteCommand.Drag(DragState.ENDED, 0.0, 0.0, MouseButtonKind.SECONDARY)),
                semanticEvent = TouchSurfaceSemanticEvent.SECONDARY_DRAG_ENDED,
            )

            else -> TouchSurfaceOutput()
        }

        resetSession()
        return output
    }

    private fun beginSession(id: Int, point: TouchPoint, timestamp: Double) {
        activeTouches[id] = ActiveTouch(startPoint = point, point = point, startTime = timestamp)
        singleTouchId = id
        sessionStartTime = timestamp
        sessionStartCentroid = point
        previousCentroid = point
        maxTouchCount = 1

        phase = if (lastTapTime?.let { timestamp - it <= doubleTapWindow } == true) {
            Phase.LEFT_DRAG_CANDIDATE
        } else {
            Phase.SINGLE_TAP_CANDIDATE
        }
    }

    private fun handleSingleMove(from: TouchPoint, to: TouchPoint): TouchSurfaceOutput {
        val dx = (to.x - from.x) * moveMultiplier
        val dy = (to.y - from.y) * moveMultiplier
        previousCentroid = to

        if (abs(dx) <= 0.05 && abs(dy) <= 0.05) return TouchSurfaceOutput()
        return TouchSurfaceOutput(commands = listOf(RemoteCommand.Move(dx, dy)))
    }

    private fun handleLeftDragCandidateMove(id: Int, point: TouchPoint): TouchSurfaceOutput {
        if (id != singleTouchId) return TouchSurfaceOutput()
        val touch = activeTouches[id] ?: return TouchSurfaceOutput()
        previousCentroid = point

        if (touch.startPoint.distanceTo(point) <= dragDistanceThreshold) return TouchSurfaceOutput()

        phase = Phase.LEFT_DRAG_ACTIVE
        val dx = (point.x - touch.startPoint.x) * moveMultiplier
        val dy = (point.y - touch.startPoint.y) * moveMultiplier
        val commands = buildList {
            add(RemoteCommand.Drag(DragState.STARTED, 0.0, 0.0, MouseButtonKind.PRIMARY))
            if (abs(dx) > 0.1 || abs(dy) > 0.1) {
                add(RemoteCommand.Drag(DragState.CHANGED, dx, dy, MouseButtonKind.PRIMARY))
            }
        }

        return TouchSurfaceOutput(
            commands = commands,
            semanticEvent = TouchSurfaceSemanticEvent.PRIMARY_DRAG_STARTED,
        )
    }

    private fun handleTwoFingerMove(timestamp: Double, oldCentroid: TouchPoint): TouchSurfaceOutput {
        if (activeTouches.size != 2) {
            phase = Phase.IDLE
            return TouchSurfaceOutput()
        }

        val newCentroid = centroid()
        previousCentroid = newCentroid
        val dx = newCentroid.x - oldCentroid.x
        val dy = newCentroid.y - oldCentroid.y
        val sessionDistance = sessionStartCentroid?.distanceTo(newCentroid) ?: 0f
        val sessionDuration = timestamp - (sessionStartTime ?: timestamp)

        if (phase == Phase.RIGHT_DRAG_ACTIVE) {
            return TouchSurfaceOutput(
                commands = listOf(
                    RemoteCommand.Drag(
                        state = DragState.CHANGED,
                        dx = dx * moveMultiplier,
                        dy = dy * moveMultiplier,
                        button = MouseButtonKind.SECONDARY,
                    ),
                ),
                semanticEvent = TouchSurfaceSemanticEvent.SECONDARY_DRAG_CHANGED,
            )
        }

        if (phase == Phase.SCROLL_ACTIVE) {
            return scrollOutput(dx = dx, dy = dy)
        }

        if (sessionDistance <= minOf(dragDistanceThreshold, scrollActivationDistance)) return TouchSurfaceOutput()

        if (sessionDuration >= rightDragHoldDelay) {
            phase = Phase.RIGHT_DRAG_ACTIVE
            val commands = buildList {
                add(RemoteCommand.Drag(DragState.STARTED, 0.0, 0.0, MouseButtonKind.SECONDARY))
                if (abs(dx) > 0.1 || abs(dy) > 0.1) {
                    add(
                        RemoteCommand.Drag(
                            state = DragState.CHANGED,
                            dx = dx * moveMultiplier,
                            dy = dy * moveMultiplier,
                            button = MouseButtonKind.SECONDARY,
                        ),
                    )
                }
            }
            return TouchSurfaceOutput(
                commands = commands,
                semanticEvent = TouchSurfaceSemanticEvent.SECONDARY_DRAG_STARTED,
            )
        }

        if (sessionDistance >= scrollActivationDistance && abs(dy) >= abs(dx) * scrollDominanceThreshold) {
            phase = Phase.SCROLL_ACTIVE
            return scrollOutput(dx = dx, dy = dy)
        }

        if (sessionDistance >= scrollActivationDistance && abs(dx) >= abs(dy) * scrollDominanceThreshold) {
            phase = Phase.SCROLL_ACTIVE
            return scrollOutput(dx = dx, dy = dy)
        }

        return TouchSurfaceOutput()
    }

    private fun scrollOutput(dx: Float, dy: Float): TouchSurfaceOutput {
        val isHorizontal = abs(dx) >= abs(dy)
        val deltaX = if (isHorizontal) -dx * scrollMultiplier else 0.0
        val deltaY = if (isHorizontal) 0.0 else -dy * scrollMultiplier
        if (abs(deltaX) <= 0.1 && abs(deltaY) <= 0.1) return TouchSurfaceOutput()
        return TouchSurfaceOutput(
            commands = listOf(RemoteCommand.Scroll(deltaX = deltaX, deltaY = deltaY)),
            semanticEvent = TouchSurfaceSemanticEvent.SCROLLING,
        )
    }

    private fun centroid(): TouchPoint {
        if (activeTouches.isEmpty()) return TouchPoint(0f, 0f)
        return TouchPoint(
            x = activeTouches.values.map { it.point.x }.average().toFloat(),
            y = activeTouches.values.map { it.point.y }.average().toFloat(),
        )
    }

    private fun dragEndOutputIfNeeded(endingTouchId: Int): TouchSurfaceOutput? =
        when {
            phase == Phase.LEFT_DRAG_ACTIVE && endingTouchId == singleTouchId -> {
                phase = Phase.IDLE
                TouchSurfaceOutput(
                    commands = listOf(RemoteCommand.Drag(DragState.ENDED, 0.0, 0.0, MouseButtonKind.PRIMARY)),
                    semanticEvent = TouchSurfaceSemanticEvent.PRIMARY_DRAG_ENDED,
                )
            }

            phase == Phase.RIGHT_DRAG_ACTIVE -> {
                phase = Phase.IDLE
                TouchSurfaceOutput(
                    commands = listOf(RemoteCommand.Drag(DragState.ENDED, 0.0, 0.0, MouseButtonKind.SECONDARY)),
                    semanticEvent = TouchSurfaceSemanticEvent.SECONDARY_DRAG_ENDED,
                )
            }

            else -> null
        }

    private fun resetIfSessionEnded() {
        if (activeTouches.isEmpty()) {
            resetSession()
        }
    }

    private fun resetSession() {
        activeTouches.clear()
        phase = Phase.IDLE
        sessionStartTime = null
        sessionStartCentroid = null
        previousCentroid = null
        maxTouchCount = 0
        singleTouchId = null
    }
}

private fun TouchPoint.distanceTo(other: TouchPoint): Float =
    hypot(other.x - x, other.y - y)

File: iOS/NeoRemote/Core/TouchSurfaceInputAdapter.swift
import CoreGraphics
import Foundation

struct TouchSurfaceInputAdapter {
    private struct ActiveTouch {
        var startPoint: CGPoint
        var point: CGPoint
        var startTime: TimeInterval
    }

    private enum Phase: Equatable {
        case idle
        case singleTapCandidate
        case leftDragCandidate
        case leftDragActive
        case multiTouchCandidate
        case scrollActive
        case rightDragActive
        case middleTapCandidate
    }

    var moveMultiplier: Double = 1
    var scrollMultiplier: Double = 1
    var tapDistanceThreshold: CGFloat = 12
    var tapDurationThreshold: TimeInterval = 0.22
    var doubleTapWindow: TimeInterval = 0.32
    var dragDistanceThreshold: CGFloat = 14
    var scrollActivationDistance: CGFloat = 14
    var rightDragHoldDelay: TimeInterval = 0.18
    var scrollDominanceThreshold: CGFloat = 1.15

    init(settings: TouchSensitivitySettings = .default) {
        apply(settings: settings)
    }

    mutating func apply(settings: TouchSensitivitySettings) {
        let clamped = settings.clamped
        moveMultiplier = clamped.cursorSensitivity
        scrollMultiplier = clamped.swipeSensitivity
        scrollActivationDistance = 14 / CGFloat(clamped.swipeSensitivity)
    }

    private var activeTouches: [Int: ActiveTouch] = [:]
    private var phase: Phase = .idle
    private var lastTapTime: TimeInterval?
    private var sessionStartTime: TimeInterval?
    private var sessionStartCentroid: CGPoint?
    private var previousCentroid: CGPoint?
    private var maxTouchCount = 0
    private var singleTouchID: Int?

    mutating func touchBegan(id: Int, point: CGPoint, timestamp: TimeInterval) -> TouchSurfaceOutput {
        if activeTouches.isEmpty {
            beginSession(id: id, point: point, timestamp: timestamp)
        } else {
            activeTouches[id] = ActiveTouch(startPoint: point, point: point, startTime: timestamp)
            maxTouchCount = max(maxTouchCount, activeTouches.count)
            sessionStartCentroid = centroid()
            previousCentroid = sessionStartCentroid
            singleTouchID = nil

            switch activeTouches.count {
            case 2:
                phase = .multiTouchCandidate
            case 3:
                phase = .middleTapCandidate
            default:
                phase = .idle
            }
        }

        return .none
    }

    mutating func touchMoved(id: Int, point: CGPoint, timestamp: TimeInterval) -> TouchSurfaceOutput {
        guard let existing = activeTouches[id] else { return .none }
        let previousPoint = existing.point
        let previousCentroid = self.previousCentroid ?? centroid()

        activeTouches[id]?.point = point

        switch phase {
        case .singleTapCandidate:
            return handleSingleMove(from: previousPoint, to: point)

        case .leftDragCandidate:
            return handleLeftDragCandidateMove(id: id, point: point)

        case .leftDragActive:
            let dx = Double(point.x - previousPoint.x) * moveMultiplier
            let dy = Double(point.y - previousPoint.y) * moveMultiplier
            return TouchSurfaceOutput(
                commands: [.drag(state: .changed, button: .primary, dx: dx, dy: dy)],
                semanticEvent: .dragChanged(.primary)
            )

        case .multiTouchCandidate, .scrollActive, .rightDragActive:
            return handleTwoFingerMove(timestamp: timestamp, previousCentroid: previousCentroid)

        case .middleTapCandidate:
            if let start = sessionStartCentroid, start.distance(to: centroid()) > tapDistanceThreshold {
                phase = .idle
            }
            return .none

        case .idle:
            return .none
        }
    }

    mutating func touchEnded(id: Int, point _: CGPoint, timestamp: TimeInterval) -> TouchSurfaceOutput {
        guard activeTouches[id] != nil else { return .none }

        let dragEnd = dragEndOutputIfNeeded(endingTouchID: id)
        activeTouches.removeValue(forKey: id)

        if let dragEnd {
            resetIfSessionEnded()
            return dragEnd
        }

        guard activeTouches.isEmpty else { return .none }
        defer { resetSession() }

        let duration = timestamp - (sessionStartTime ?? timestamp)
        let movement = sessionStartCentroid.map { $0.distance(to: previousCentroid ?? $0) } ?? 0

        switch phase {
        case .singleTapCandidate, .leftDragCandidate:
            guard duration <= tapDurationThreshold, movement <= tapDistanceThreshold else { return .none }
            lastTapTime = timestamp
            return TouchSurfaceOutput(commands: [.tap(kind: .primary)], semanticEvent: .tap(.primary))

        case .multiTouchCandidate:
            guard maxTouchCount == 2, duration <= tapDurationThreshold, movement <= tapDistanceThreshold else { return .none }
            return TouchSurfaceOutput(commands: [.tap(kind: .secondary)], semanticEvent: .tap(.secondary))

        case .middleTapCandidate:
            guard maxTouchCount == 3, duration <= tapDurationThreshold, movement <= tapDistanceThreshold else { return .none }
            return TouchSurfaceOutput(commands: [.tap(kind: .middle)], semanticEvent: .tap(.middle))

        default:
            return .none
        }
    }

    mutating func cancelAllTouches() -> TouchSurfaceOutput {
        defer { resetSession() }

        switch phase {
        case .leftDragActive:
            return TouchSurfaceOutput(
                commands: [.drag(state: .ended, button: .primary, dx: 0, dy: 0)],
                semanticEvent: .dragEnded(.primary)
            )
        case .rightDragActive:
            return TouchSurfaceOutput(
                commands: [.drag(state: .ended, button: .secondary, dx: 0, dy: 0)],
                semanticEvent: .dragEnded(.secondary)
            )
        default:
            return .none
        }
    }

    private mutating func beginSession(id: Int, point: CGPoint, timestamp: TimeInterval) {
        activeTouches[id] = ActiveTouch(startPoint: point, point: point, startTime: timestamp)
        singleTouchID = id
        sessionStartTime = timestamp
        sessionStartCentroid = point
        previousCentroid = point
        maxTouchCount = 1

        if let lastTapTime, timestamp - lastTapTime <= doubleTapWindow {
            phase = .leftDragCandidate
        } else {
            phase = .singleTapCandidate
        }
    }

    private mutating func handleSingleMove(from previousPoint: CGPoint, to point: CGPoint) -> TouchSurfaceOutput {
        let dx = Double(point.x - previousPoint.x) * moveMultiplier
        let dy = Double(point.y - previousPoint.y) * moveMultiplier
        previousCentroid = point

        guard abs(dx) > 0.05 || abs(dy) > 0.05 else { return .none }
        return TouchSurfaceOutput(commands: [.move(dx: dx, dy: dy)], semanticEvent: nil)
    }

    private mutating func handleLeftDragCandidateMove(id: Int, point: CGPoint) -> TouchSurfaceOutput {
        guard id == singleTouchID, let touch = activeTouches[id] else { return .none }
        previousCentroid = point
        let distance = touch.startPoint.distance(to: point)

        guard distance > dragDistanceThreshold else { return .none }

        phase = .leftDragActive
        let dx = Double(point.x - touch.startPoint.x) * moveMultiplier
        let dy = Double(point.y - touch.startPoint.y) * moveMultiplier
        var commands: [RemoteCommand] = [.drag(state: .started, button: .primary, dx: 0, dy: 0)]
        if abs(dx) > 0.1 || abs(dy) > 0.1 {
            commands.append(.drag(state: .changed, button: .primary, dx: dx, dy: dy))
        }
        return TouchSurfaceOutput(commands: commands, semanticEvent: .dragStarted(.primary))
    }

    private mutating func handleTwoFingerMove(
        timestamp: TimeInterval,
        previousCentroid: CGPoint
    ) -> TouchSurfaceOutput {
        guard activeTouches.count == 2 else {
            phase = .idle
            return .none
        }

        let newCentroid = centroid()
        self.previousCentroid = newCentroid
        let dx = newCentroid.x - previousCentroid.x
        let dy = newCentroid.y - previousCentroid.y
        let sessionDistance = sessionStartCentroid.map { $0.distance(to: newCentroid) } ?? 0
        let sessionDuration = timestamp - (sessionStartTime ?? timestamp)

        if phase == .rightDragActive {
            return TouchSurfaceOutput(
                commands: [.drag(state: .changed, button: .secondary, dx: Double(dx) * moveMultiplier, dy: Double(dy) * moveMultiplier)],
                semanticEvent: .dragChanged(.secondary)
            )
        }

        if phase == .scrollActive {
            return scrollOutput(dx: dx, dy: dy)
        }

        guard sessionDistance > min(dragDistanceThreshold, scrollActivationDistance) else { return .none }

        if sessionDuration >= rightDragHoldDelay {
            phase = .rightDragActive
            var commands: [RemoteCommand] = [.drag(state: .started, button: .secondary, dx: 0, dy: 0)]
            if abs(dx) > 0.1 || abs(dy) > 0.1 {
                commands.append(.drag(state: .changed, button: .secondary, dx: Double(dx) * moveMultiplier, dy: Double(dy) * moveMultiplier))
            }
            return TouchSurfaceOutput(commands: commands, semanticEvent: .dragStarted(.secondary))
        }

        if sessionDistance >= scrollActivationDistance, abs(dy) >= abs(dx) * scrollDominanceThreshold {
            phase = .scrollActive
            return scrollOutput(dx: dx, dy: dy)
        }

        if sessionDistance >= scrollActivationDistance, abs(dx) >= abs(dy) * scrollDominanceThreshold {
            phase = .scrollActive
            return scrollOutput(dx: dx, dy: dy)
        }

        return .none
    }

    private func scrollOutput(dx: CGFloat, dy: CGFloat) -> TouchSurfaceOutput {
        let isHorizontal = abs(dx) >= abs(dy)
        let deltaX = isHorizontal ? Double(-dx) * scrollMultiplier : 0
        let deltaY = isHorizontal ? 0 : Double(-dy) * scrollMultiplier
        guard abs(deltaX) > 0.1 || abs(deltaY) > 0.1 else { return .none }
        return TouchSurfaceOutput(commands: [.scroll(deltaX: deltaX, deltaY: deltaY)], semanticEvent: .scrolling)
    }

    private func centroid() -> CGPoint {
        guard !activeTouches.isEmpty else { return .zero }
        let x = activeTouches.values.map(\.point.x).reduce(0, +) / CGFloat(activeTouches.count)
        let y = activeTouches.values.map(\.point.y).reduce(0, +) / CGFloat(activeTouches.count)
        return CGPoint(x: x, y: y)
    }

    private mutating func dragEndOutputIfNeeded(endingTouchID id: Int) -> TouchSurfaceOutput? {
        switch phase {
        case .leftDragActive where id == singleTouchID:
            phase = .idle
            return TouchSurfaceOutput(
                commands: [.drag(state: .ended, button: .primary, dx: 0, dy: 0)],
                semanticEvent: .dragEnded(.primary)
            )

        case .rightDragActive:
            phase = .idle
            return TouchSurfaceOutput(
                commands: [.drag(state: .ended, button: .secondary, dx: 0, dy: 0)],
                semanticEvent: .dragEnded(.secondary)
            )

        default:
            return nil
        }
    }

    private mutating func resetIfSessionEnded() {
        if activeTouches.isEmpty {
            resetSession()
        }
    }

    private mutating func resetSession() {
        activeTouches.removeAll()
        phase = .idle
        sessionStartTime = nil
        sessionStartCentroid = nil
        previousCentroid = nil
        maxTouchCount = 0
        singleTouchID = nil
    }
}
